### PR TITLE
Use ClientErrorData to configure ClientErrorResultFilter

### LINKED ISSUE #8289: Review the extensibility for configuring client error factories
The API for configuring client error factories needs a bunch of polish.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #8289

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @khellang

File: src/Microsoft.AspNetCore.Mvc.Core/ApiBehaviorOptions.cs [hunk 1]
```diff
@@ -71,12 +71,12 @@ public ApiBehaviorOptions()
...
-        /// to transform certain certain client errors.
+        /// Gets or sets a value that determines if controllers with <see cref="ApiControllerAttribute"/>
+        /// transform certain certain client errors.
         /// <para>
-        /// When <c>false</c>, <see cref="ClientErrorFactory"/> is used to transform <see cref="IClientErrorActionResult"/> to the value
-        /// specified by the factory. In the default case, this converts <see cref="StatusCodeResult"/> instances to an <see cref="ObjectResult"/>
-        /// with <see cref="ProblemDetails"/>.
+        /// When <c>false</c>, a result filter is added to API controller actions that transforms <see cref="IClientErrorActionResult"/>.
+        /// By default, <see cref="ClientErrorMapping"/> is used to map <see cref="IClientErrorActionResult"/> to a
+        /// <see cref="ProblemDetails"/> instance (returned as the value for <see cref="ObjectResult"/>).
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ApiBehaviorOptions.cs]: This should explain how to replace the service if you want to use something other than problem details.

File: src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/ClientErrorResultFilter.cs [hunk 2]
```diff
@@ -38,16 +37,15 @@ public void OnResultExecuting(ResultExecutingContext context)
...
-                clientErrorActionResult.StatusCode is int statusCode &&
-                _clientErrorFactory.TryGetValue(statusCode, out var factory))
+            if (!(context.Result is IClientErrorActionResult clientError))
             {
-                var result = factory(context);
+                return;
+            }
 
-                _logger.TransformingClientError(context.Result.GetType(), result?.GetType(), statusCode);
+            var result = _clientErrorFactory.GetClientError(context, clientError);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/ClientErrorResultFilter.cs]: This should be allowed to return null? It makes sense that we'd want to allow users to write code that decides **not** to map something, in which case we'd use the original error.

File: src/Microsoft.AspNetCore.Mvc.Core/ClientErrorData.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        public string Link { get; set; }
+
+        /// <summary>
+        /// Gets or sets the summary of the client error.
+        /// </summary>
+        /// <remarks>
+        /// By default, this maps to <see cref="ProblemDetails.Title"/> and should not change
+        /// between multiple occurences of the same error.
+        /// </remarks>
+        public string Summary { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ClientErrorData.cs]: Should this be `Title`? I feel like title is a neutral enough term.

**Comment by pranavkm** [comment]
🆙 📅

**Pull request merged**